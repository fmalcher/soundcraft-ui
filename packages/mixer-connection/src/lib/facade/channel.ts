import { take } from 'rxjs/operators';
import { MixerConnection } from '../mixer-connection';
import { MixerStore } from '../state/mixer-store';
import {
  select,
  selectFaderValue,
  selectMute,
  selectPan,
} from '../state/state-selectors';
import { BusType, ChannelType } from '../types';

/**
 * Represents a single channel with a fader
 */
export class Channel {
  protected fullChannelId = `${this.channelType}.${this.channel - 1}`;
  protected faderLevelCommand = 'mix';

  faderLevel$ = this.store.state$.pipe(
    select(
      selectFaderValue(this.channelType, this.channel, this.busType, this.bus)
    )
  );

  pan$ = this.store.state$.pipe(
    select(selectPan(this.channelType, this.channel, this.busType, this.bus))
  );

  mute$ = this.store.state$.pipe(
    select(selectMute(this.channelType, this.channel, this.busType, this.bus))
  );

  constructor(
    protected conn: MixerConnection,
    protected store: MixerStore,
    protected channelType: ChannelType,
    protected channel: number,
    protected busType: BusType = 'master',
    protected bus: number = 0
  ) {}

  setFaderLevel(value: number) {
    const command = `SETD^${this.fullChannelId}.${this.faderLevelCommand}^${value}`;
    this.conn.sendMessage(command);
  }

  setMute(value: number) {
    const command = `SETD^${this.fullChannelId}.mute^${value}`;
    this.conn.sendMessage(command);
  }

  mute() {
    this.setMute(1);
  }

  unmute() {
    this.setMute(0);
  }

  toggleMute() {
    this.mute$.pipe(take(1)).subscribe(mute => this.setMute(mute ^ 1));
  }
}
