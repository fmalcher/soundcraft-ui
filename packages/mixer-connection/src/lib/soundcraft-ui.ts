import { delay, filter, take } from 'rxjs/operators';
import { AuxBus } from './facade/aux-bus';
import { FxBus } from './facade/fx-bus';
import { MasterBus } from './facade/master-bus';
import { MuteGroup, MuteGroupID } from './facade/mute-group';
import { Player } from './facade/player';
import { MixerConnection } from './mixer-connection';
import { MixerStore } from './state/mixer-store';
import { ConnectionStatus } from './types';

export class SoundcraftUI {
  readonly conn = new MixerConnection(this.targetIP);
  readonly store = new MixerStore(this.conn);

  /** Connection status */
  status$ = this.conn.status$;

  /** Get master bus */
  master = new MasterBus(this.conn, this.store);

  /** Get media player */
  player = new Player(this.conn, this.store);

  constructor(private targetIP: string) {}

  /**
   * Get AUX bus
   * @param bus Bus number
   */
  aux(bus: number) {
    return new AuxBus(this.conn, this.store, bus);
  }

  /**
   * Get FX bus
   * @param bus Bus number
   */
  fx(bus: number) {
    return new FxBus(this.conn, this.store, bus);
  }

  /**
   * Get MUTE group or related groupings (MUTE ALL and MUTE FX)
   * @param id ID of the group: 1..6, all, fx
   */
  muteGroup(id: MuteGroupID) {
    return new MuteGroup(this.conn, this.store, id);
  }

  /** Unmute all mute groups, "MUTE ALL" and "MUTE FX" */
  clearMuteGroups() {
    this.conn.sendMessage('SETD^mgmask^0');
  }

  /** Connect to the mixer */
  connect() {
    this.conn.connect();
  }

  /** Disconnect from the mixer */
  disconnect() {
    this.conn.disconnect();
  }

  /**
   * Reconnect to the mixer:
   * disconnect, then wait 1 second before connecting again
   */
  reconnect() {
    this.conn.status$
      .pipe(
        filter(e => e.type === ConnectionStatus.Close),
        take(1),
        delay(1000)
      )
      .subscribe(() => this.conn.connect());

    this.conn.disconnect();
  }
}
