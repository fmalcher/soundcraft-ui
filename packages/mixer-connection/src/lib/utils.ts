import { debounceTime, firstValueFrom, race, take, timer } from 'rxjs';
import { MixerStore } from './state/mixer-store';
import { ChannelType, FxType } from './types';

/** Clamp numeric value to min and max */
export function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

/**
 * Transform a given value to int, float or string
 * @param value
 */
export function transformStringValue(value: string) {
  if (value.match(/^-?\d+$/)) {
    return parseInt(value, 10);
  } else if (value.match(/^\d+\.\d+$/)) {
    return parseFloat(value);
  } else {
    return value;
  }
}

/**
 * Transform player time in seconds to human-readable format M:SS
 * @param value player time in seconds
 */
export function playerTimeToString(value: number) {
  if (value < 0) {
    return '';
  }
  const cleanValue = Math.floor(value);
  const minutes = Math.floor(cleanValue / 60);
  const seconds = cleanValue % 60;
  return `${minutes}:${seconds.toString().length === 1 ? '0' : ''}${seconds}`;
}

export function getLinkedChannelNumber(channel: number, stereoIndex: number): number {
  switch (stereoIndex) {
    case 1:
      return channel - 1;
    case 0:
      return channel + 1;
    default:
      return channel;
  }
}

/** Helper function to convert channel number into readable L/R value */
function numberToLR(channel: number): string {
  switch (channel) {
    case 1:
      return 'L';
    case 2:
      return 'R';
    default:
      return '';
  }
}

/** Helper function to convert FX type into readable name (Reverb, Chorus, ...) */
export function fxTypeToString(type: FxType): keyof typeof FxType {
  switch (type) {
    case FxType.None:
      return 'None';
    case FxType.Chorus:
      return 'Chorus';
    case FxType.Reverb:
      return 'Reverb';
    case FxType.Delay:
      return 'Delay';
    case FxType.Room:
      return 'Room';
  }
}

/**
 * Construct a human-readable name for a channel
 * based on the default labels from the web interface
 * @param type
 * @param channel
 * @returns
 */
export function constructReadableChannelName(type: ChannelType, channel: number): string {
  switch (type) {
    case 'i':
      return 'CH ' + channel;
    case 'a':
      return 'AUX ' + channel;
    case 'f':
      return 'FX ' + channel;
    case 's':
      return 'SUB ' + channel;
    case 'v':
      return 'VCA ' + channel;
    case 'l':
      return 'LINE IN ' + numberToLR(channel);
    case 'p':
      return 'PLAYER ' + numberToLR(channel);
  }
}

/**
 * Returns a Promise that fires when the mixer state hasn't changed for 25 ms OR when 250 ms timeout are over.
 * This makes sure that all initial params can be received by the mixer after connection init.
 * In case the state never stands still for 50 ms, the 250 ms timeout will emit finally.
 */
export function waitForInitParams(store: MixerStore): Promise<void> {
  return firstValueFrom(race(store.state$.pipe(debounceTime(25)), timer(250)));
}
