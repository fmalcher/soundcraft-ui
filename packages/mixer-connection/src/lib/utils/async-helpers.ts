/**
 * Returns a promise that resolves delayed by `delayMs` milliseconds
 */
export function resolveDelayed(delayMs: number): Promise<void> {
  return new Promise<void>(resolve => setTimeout(() => resolve(), delayMs));
}
