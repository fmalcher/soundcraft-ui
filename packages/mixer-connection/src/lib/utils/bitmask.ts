/**
 * Toggle a bit of `value`
 * @param value Source value
 * @param bitIndex Bit index, right to left
 */
export function toggleBit(value: number, bitIndex: number) {
  if (bitIndex < 0) {
    return value;
  }
  return value ^ (1 << bitIndex);
}

/**
 * Change a bit of `value` to 0
 * @param value Source value
 * @param bitIndex Bit index, right to left
 */
export function clearBit(value: number, bitIndex: number) {
  if (bitIndex < 0) {
    return value;
  }
  return value & ~(1 << bitIndex);
}

/**
 * Change a bit of `value` to 1
 * @param value Source value
 * @param bitIndex Bit index, right to left
 */
export function setBit(value: number, bitIndex: number) {
  if (bitIndex < 0) {
    return value;
  }
  return value | (1 << bitIndex);
}

/**
 * Return the value of a specific bit in `value`
 * @param value Source value
 * @param bitIndex Bit index, right to left
 */
export function getValueOfBit(value: number, bitIndex: number) {
  if (bitIndex < 0) {
    return value;
  }
  return value & (1 << bitIndex) ? 1 : 0;
}
