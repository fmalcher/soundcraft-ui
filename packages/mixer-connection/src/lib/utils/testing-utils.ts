import { Observable } from 'rxjs';
import { first } from 'rxjs/operators';

export function readFirst<T>(o: Observable<T>): Promise<T> {
  return o.pipe(first()).toPromise();
}
