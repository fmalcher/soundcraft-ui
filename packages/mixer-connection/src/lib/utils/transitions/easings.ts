export enum Easings {
  Linear,
  EaseIn,
  EaseOut,
  EaseInOut,
}

export const easingFunctions: { [key in Easings]: (t: number) => number } = {
  [Easings.Linear]: (t: number) => t,
  [Easings.EaseIn]: (t: number) => t * t,
  [Easings.EaseOut]: (t: number) => t * (2 - t),
  [Easings.EaseInOut]: (t: number) => t * (2 - t),
};
