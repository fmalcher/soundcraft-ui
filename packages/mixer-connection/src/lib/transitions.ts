import { interval, Observable, Subject, merge } from 'rxjs';
import {
  distinctUntilChanged,
  filter,
  groupBy,
  map,
  mergeMap,
  switchMap,
  take,
  takeUntil,
} from 'rxjs/operators';
import { MixerConnection } from './mixer-connection';
import { ConnectionStatus } from './types';
import { easingFunctions, Easings } from './utils/transitions/easings';

export interface TransitionSource {
  fullChannelId: string;
  faderLevelCommand: string;
  sourceValue: number;
  targetValue: number;
  fadeTime: number;
  easing: Easings;
  fps: number;
}

export class TransitionRegistry {
  private sources$ = new Subject<TransitionSource>();
  private stopTransitions$ = new Subject();

  constructor(private conn: MixerConnection) {
    /**
     * Group the transition sources by channel. This allows us to cancel a running transition on a specific channel when a new one appears (using switchMap).
     * All other running transitions are not affected by this.
     */
    this.sources$
      .pipe(
        groupBy(s => s.fullChannelId),
        mergeMap(sourcesPerChannel$ =>
          sourcesPerChannel$.pipe(
            switchMap(source =>
              this.generateTransition(source).pipe(
                map(value =>
                  this.buildLevelCommand(
                    source.fullChannelId,
                    source.faderLevelCommand,
                    value
                  )
                )
              )
            )
          )
        ),
        // stop all transitions when notifier calls or when connection is closed
        takeUntil(
          merge(
            this.stopTransitions$,
            this.conn.status$.pipe(
              filter(e => e.type === ConnectionStatus.Close)
            )
          )
        )
      )
      .subscribe(command => this.conn.sendMessage(command));
  }

  /**
   * Add a new transition to perform
   * @param source A TransitionSource object with all metadata about the transition
   */
  addTransition(source: TransitionSource) {
    this.sources$.next(source);
  }

  /** Stop all running transitions */
  stopAll() {
    this.stopTransitions$.next(null);
  }

  /** Generate command to set the fader level */
  private buildLevelCommand(
    fullChannelId: string,
    faderLevelCommand: string,
    value: number
  ) {
    return `SETD^${fullChannelId}.${faderLevelCommand}^${value}`;
  }

  /**
   * Perform the transition by generating the discrete steps
   * @param source TransitionSource object with metadata about the transition
   */
  private generateTransition({
    sourceValue,
    targetValue,
    fadeTime,
    easing,
    fps,
  }: TransitionSource): Observable<number> {
    fadeTime = Math.max(fadeTime, fps); // fade time must not be smaller than fps
    const stepTime = Math.round(1000 / fps);
    const steps = Math.round(fadeTime / stepTime);

    easing = easing || Easings.Linear;
    const easingFn = easingFunctions[easing] || easingFunctions[Easings.Linear];

    return interval(stepTime).pipe(
      take(steps),
      map(
        i =>
          easingFn((i + 1) / steps) * (targetValue - sourceValue) + sourceValue
      ),
      distinctUntilChanged()
    );
  }
}
