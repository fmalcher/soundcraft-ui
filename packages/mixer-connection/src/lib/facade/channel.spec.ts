import { SoundcraftUI } from '../soundcraft-ui';
import { readFirst } from '../utils/testing-utils';
import { Easings } from '../utils/transitions/easings';
import { Channel } from './channel';

describe('Channel', () => {
  let conn: SoundcraftUI;
  let channel: Channel;

  beforeEach(() => {
    conn = new SoundcraftUI('0.0.0.0');
    channel = conn.master.input(2);
  });

  describe('Fader Level', () => {
    it('faderLevel$', async () => {
      channel.setFaderLevel(0.5);
      expect(await readFirst(channel.faderLevel$)).toBe(0.5);

      channel.setFaderLevel(0.8889);
      expect(await readFirst(channel.faderLevel$)).toBe(0.8889);

      channel.setFaderLevelDB(0);
      expect(await readFirst(channel.faderLevel$)).toBe(0.76470588235);
    });

    it('faderLevelDB$', async () => {
      channel.setFaderLevelDB(-6);
      expect(await readFirst(channel.faderLevelDB$)).toBe(-6);

      channel.setFaderLevelDB(-12);
      expect(await readFirst(channel.faderLevelDB$)).toBe(-12);

      channel.setFaderLevel(0.76470588235);
      expect(await readFirst(channel.faderLevelDB$)).toBe(0);
    });

    it('changeFaderLevelDB', async () => {
      channel.setFaderLevelDB(-12);
      channel.changeFaderLevelDB(3);
      expect(await readFirst(channel.faderLevelDB$)).toBe(-9);

      channel.changeFaderLevelDB(-6);
      expect(await readFirst(channel.faderLevelDB$)).toBe(-15);
    });
  });

  describe('MUTE', () => {
    it('mute$', async () => {
      channel.mute();
      expect(await readFirst(channel.mute$)).toBe(1);

      channel.unmute();
      expect(await readFirst(channel.mute$)).toBe(0);

      channel.setMute(1);
      expect(await readFirst(channel.mute$)).toBe(1);
    });

    it('toggleMute', async () => {
      channel.setMute(0);
      channel.toggleMute();
      expect(await readFirst(channel.mute$)).toBe(1);

      channel.toggleMute();
      expect(await readFirst(channel.mute$)).toBe(0);
    });
  });

  describe('Fades', () => {
    let results: number[];

    beforeEach(() => {
      jest.useFakeTimers();
      results = [];
    });

    afterEach(() => jest.useRealTimers());

    describe('fadeTo', () => {
      beforeEach(() => {
        channel.faderLevel$.subscribe(e => results.push(e));
      });

      it('Linear', () => {
        channel.setFaderLevel(0.3);
        channel.fadeTo(0.8, 1000, Easings.Linear);
        jest.advanceTimersByTime(1000);
        expect(results).toMatchSnapshot();
      });

      it('EaseIn', () => {
        channel.setFaderLevel(1);
        channel.fadeTo(0.2, 1000, Easings.EaseIn);
        jest.advanceTimersByTime(1000);
        expect(results).toMatchSnapshot();
      });

      it('EaseOut', () => {
        channel.setFaderLevel(0.7);
        channel.fadeTo(0.9, 1000, Easings.EaseOut);
        jest.advanceTimersByTime(1000);
        expect(results).toMatchSnapshot();
      });

      it('EaseInOut', () => {
        channel.setFaderLevel(0.1);
        channel.fadeTo(0.4, 1000, Easings.EaseInOut);
        jest.advanceTimersByTime(1000);
        expect(results).toMatchSnapshot();
      });
    });

    describe('fadeToDB', () => {
      beforeEach(() => {
        channel.faderLevelDB$.subscribe(e => results.push(e));
      });

      it('Linear', () => {
        channel.setFaderLevelDB(-60);
        channel.fadeToDB(-3, 1000, Easings.Linear);
        jest.advanceTimersByTime(1000);
        expect(results).toMatchSnapshot();
      });

      it('EaseIn', () => {
        channel.setFaderLevelDB(3);
        channel.fadeToDB(-15, 1000, Easings.EaseIn);
        jest.advanceTimersByTime(1000);
        expect(results).toMatchSnapshot();
      });

      it('EaseOut', () => {
        channel.setFaderLevelDB(3);
        channel.fadeToDB(-15, 1000, Easings.EaseOut);
        jest.advanceTimersByTime(1000);
        expect(results).toMatchSnapshot();
      });

      it('EaseInOut', () => {
        channel.setFaderLevelDB(6);
        channel.fadeToDB(-6, 1000, Easings.EaseInOut);
        jest.advanceTimersByTime(1000);
        expect(results).toMatchSnapshot();
      });
    });
  });
});
