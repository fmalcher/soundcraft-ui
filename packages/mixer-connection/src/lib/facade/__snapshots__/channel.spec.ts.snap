// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Channel Fades fadeTo EaseIn 1`] = `
Array [
  1,
  0.99872,
  0.99488,
  0.98848,
  0.97952,
  0.968,
  0.95392,
  0.93728,
  0.91808,
  0.89632,
  0.872,
  0.84512,
  0.81568,
  0.7836799999999999,
  0.74912,
  0.712,
  0.67232,
  0.63008,
  0.58528,
  0.53792,
  0.4879999999999999,
  0.43552,
  0.38047999999999993,
  0.32287999999999994,
  0.26271999999999995,
  0.19999999999999996,
]
`;

exports[`Channel Fades fadeTo EaseInOut 1`] = `
Array [
  0.1,
  0.12352,
  0.14608000000000002,
  0.16768,
  0.18832000000000002,
  0.20800000000000002,
  0.22672000000000003,
  0.24448000000000003,
  0.26128,
  0.27712000000000003,
  0.29200000000000004,
  0.30592,
  0.31888000000000005,
  0.33088000000000006,
  0.34192,
  0.352,
  0.36112,
  0.36928000000000005,
  0.37648000000000004,
  0.38272000000000006,
  0.388,
  0.3923200000000001,
  0.39568000000000014,
  0.3980800000000001,
  0.39952,
  0.4,
]
`;

exports[`Channel Fades fadeTo EaseOut 1`] = `
Array [
  0.7,
  0.71568,
  0.7307199999999999,
  0.74512,
  0.75888,
  0.772,
  0.78448,
  0.79632,
  0.80752,
  0.81808,
  0.8280000000000001,
  0.83728,
  0.84592,
  0.85392,
  0.86128,
  0.868,
  0.87408,
  0.87952,
  0.88432,
  0.88848,
  0.892,
  0.89488,
  0.89712,
  0.8987200000000001,
  0.89968,
  0.9,
]
`;

exports[`Channel Fades fadeTo Linear 1`] = `
Array [
  0.3,
  0.32,
  0.33999999999999997,
  0.36,
  0.38,
  0.4,
  0.42,
  0.44,
  0.45999999999999996,
  0.48,
  0.5,
  0.52,
  0.54,
  0.56,
  0.5800000000000001,
  0.6,
  0.62,
  0.64,
  0.6599999999999999,
  0.6799999999999999,
  0.7,
  0.72,
  0.74,
  0.76,
  0.78,
  0.8,
]
`;

exports[`Channel Fades fadeToDB EaseIn 1`] = `
Array [
  3,
  3,
  2.9,
  2.8,
  2.6,
  2.4,
  2.1,
  1.8,
  1.4,
  1,
  0.6,
  0.1,
  -0.5,
  -1.1,
  -1.7,
  -2.4,
  -3.2,
  -4,
  -5,
  -6,
  -7.1,
  -8.3,
  -9.7,
  -11.3,
  -13,
  -15,
]
`;

exports[`Channel Fades fadeToDB EaseInOut 1`] = `
Array [
  6,
  5,
  4,
  3.2,
  2.4,
  1.6,
  0.9,
  0.2,
  -0.4,
  -1,
  -1.6,
  -2.1,
  -2.6,
  -3.1,
  -3.6,
  -4,
  -4.3,
  -4.7,
  -5,
  -5.2,
  -5.5,
  -5.7,
  -5.8,
  -5.9,
  -6,
  -6,
]
`;

exports[`Channel Fades fadeToDB EaseOut 1`] = `
Array [
  3,
  1.8,
  0.7,
  -0.4,
  -1.4,
  -2.4,
  -3.4,
  -4.4,
  -5.3,
  -6.2,
  -7.1,
  -8,
  -8.8,
  -9.6,
  -10.4,
  -11.1,
  -11.8,
  -12.4,
  -13,
  -13.5,
  -14,
  -14.3,
  -14.6,
  -14.8,
  -15,
  -15,
]
`;

exports[`Channel Fades fadeToDB Linear 1`] = `
Array [
  -60,
  -55.6,
  -51.4,
  -47.5,
  -43.8,
  -40.3,
  -37,
  -34,
  -31.1,
  -28.4,
  -25.9,
  -23.6,
  -21.4,
  -19.3,
  -17.4,
  -15.7,
  -14,
  -12.5,
  -11,
  -9.7,
  -8.4,
  -7.2,
  -6.1,
  -5,
  -4,
  -3,
]
`;
